src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1711,29 +1807,70 @@ RelationInfoContainsReferenceTable(PlannerInfo *plannerInfo, RelOptInfo *relatio
 
 
 /*
- * HasReferenceTable checks whether there exist a reference table in the
- * given node.
+ * HasRecurringTuples returns whether any part of the expression
+ * is recurring in every shard query.
  */
 bool
-HasReferenceTable(Node *node)
+HasRecurringTuples(Node *node, RecurringTuplesType *recurType)
 {
-	List *relationList = NIL;
-	ListCell *relationCell = NULL;
-	ExtractRangeTableRelationWalkerWithRTEExpand(node, &relationList);
+	if (node == NULL)
+	{
+		return false;
+	}
 
-	foreach(relationCell, relationList)
+	if (IsA(node, RangeTblEntry))
 	{
-		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(relationCell);
-		Oid relationId = rangeTableEntry->relid;
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) node;
 
-		if (IsDistributedTable(relationId) && PartitionMethod(relationId) ==
-			DISTRIBUTE_BY_NONE)
+		if (rangeTableEntry->rtekind == RTE_RELATION)
 		{
+			Oid relationId = rangeTableEntry->relid;
+			if (IsDistributedTable(relationId) &&
+				PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+			{
+				*recurType = RECURRING_TUPLES_REFERENCE_TABLE;
+
+				/*
+				 * Tuples from reference tables will recur in every shard query
+				 * that includes it.
+				 */
+				return true;
+			}
+		}
+		else if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			*recurType = RECURRING_TUPLES_FUNCTION;
+
+			/*
+			 * Tuples from functions will recur in every shard query that includes
+			 * it.
+			 */
 			return true;
 		}
+
+		return range_table_walker(list_make1(rangeTableEntry), HasRecurringTuples,
```

Comment:
Hmm, don't we need to pass `QTW_EXAMINE_RTES` flag?  I remember we discussed that with @velioglu while he was working on similar codes, but, I forgot the reason we skipped that :/

Reply:
then we'd recurse back into this function infinitely.